### ISSUE TITLE
discriminated union patterns with better error messages than M.or()

### ISSUE THREAD

**Issue by dckc** [Agoric/Agoric#6115]
## What is the Problem Being Solved?

We can write a pattern for a discriminated union using `M.or()`, but when a `fit()` match fails, the diagnostic isn't specialized to any branch of the union.

## Description of the Design

Where we can now write...

```js
M.or(
  {kind: 'animal', eats: FoodShape},
  {kind: 'vegetable', lifetime: M.number()},
  ...
);
```

if we could write...

```js
M.choose('kind', [
  {kind: 'animal', eats: FoodShape},
  {kind: 'vegetable', lifetime: M.number()},
  ...
])
```

Then perhaps a diagnostic for `{kind: 'animal', lifetime: 3}` could say "property eats missing" rather than just "didn't match".

## Security Considerations

I suppose the diagnostic reveals a bit more information, but the pattern writer can choose whether that's appropriate.

## Test Plan

normal unit tests should suffice

cc @erights @Chris-Hibbert @gibson042

**Comment by kriskowal** [comment]
On the off chance it’s useful, JSON Schema has some precedents: https://json-schema.org/understanding-json-schema/reference/combining.html

`anyOf`, `allOf`, `oneOf`, & `not`. I found that this set of operators was sufficient to implement an algebraic simplifier out of which it was straightforward to implement a backward-compatibility detector (does schema A extend schema B?), which I’m sure we’ll inevitably need.

**Comment by erights** [comment]
> implement a backward-compatibility detector (does schema A extend schema B?), which I’m sure we’ll inevitably need.

Neat!

**Comment by erights** [comment]
Their `allOf` is our `M.and()`
Their `anyOf` is our `M.or()`
Their `not` is our `M.not()`

We have no equivalent to `oneOf`. We could if it is useful enough, but I am skeptical that it is. I suspect that almost all uses of their `oneOf` could be replaced with `anyOf` without losing correctness. Correctness is only lost if in the original:
   * the subpatterns themselves are not disjoint
   * a specimen falling in the intersection of the subpatterns should be rejected

Any idea how often this has been the case, and whether these cases are compelling? Even for those, would they be better expressed anyway by making the subpatterns disjoint? I'd need to see some examples.

**Comment by erights** [comment]
I like the form of `M.choose()` that @dckc describes at https://github.com/Agoric/agoric-sdk/issues/6115#issue-1360472898 . It is very cool that its semantics is exactly an `M.or()` ignoring the first argument, and that the difference is only that it generates better diagnostics when it fails.

**Comment by kriskowal** [comment]
Yes, and anyone who truly needs mutual exclusion can construct `xor` from (`not`, `and`, and `or`) as foretold by the lawyer DeMorgan.

**Comment by kriskowal** [comment]
(This is in fact the secret sauce of the backward compatibility checker: It reduces all schemas to a “barrel of NAND” and co-visits the resulting schemas looking for any contradiction.)

**Comment by erights** [comment]
It is interesting that an algebraic pattern query optimizer, that matches or doesn't match exactly when the original does, can be much more effective than one that also preserves the error messages when it doesn't match. OTOH, given that we only care about matching cost when it matches, we can keep the original around just to generate the original error message when the optimized pattern does not match. Weird!

**Comment by aj-agoric** [comment]
movign this from ready to backlog due to activity, please lmk if I am out of line here

**Comment by erights** [comment]
What activity?

**Comment by gibson042** [comment]
Moved to Endo: https://github.com/endojs/endo/issues/3072
